sdk-core-protos/protos/local/temporal/sdk/core/workflow_activation/workflow_activation.proto

```diff
@@ -119,6 +121,12 @@ message WorkflowActivationJob {
         ResolveRequestCancelExternalWorkflow resolve_request_cancel_external_workflow = 13;
         // A request to handle a workflow update.
         DoUpdate do_update = 14;
+        // A nexus operation started.
+        ResolveNexusOperationStart resolve_nexus_operation_start = 15;
+        // A nexus operation resolved.
+        ResolveNexusOperation resolve_nexus_operation = 16;
+        // A request to cancel a nexus operation resolved.
+        ResolveCancelNexusOperation resolve_cancel_nexus_operation = 17;
```

Comment:
Hrmm, I guess this doesn't mean a Nexus operation was cancelled, but rather that the cancel was sent? I do wonder what situation we need this as an awaitable thing that can come back. I would similarly not expect a `ResolveCancelChildWorkflowExecution` since it effectively is always expected to succeed. But I do see the `NexusOperationCancelRequestedEventAttributes` event. Hrmm.

The issue from a lang POV is that cancellation of things you start often does not have a path for "resolving" of the cancellation alone. Because, unlike external workflow cancellation, we often use cancellation token like things when starting things, we don't expose a fallible/awaitable cancellation. If we supported cancelling Nexus operations you didn't start, definitely would be worth the extra wait.

Reply:
They need to be awaitable. Support will be added later for ensuring cancel was received, so we need to have them be awaitable. This is a lot like how child workflows get cancelled via the external workflow cancel command.

And yeah, cancelling ones you didn't start likely will be supported at some point too.

Reply:
As for the cancel token thing, yeah, in this case you're just going to end up internally seeing the resolution and maybe not showing it to the user. But, likely the type that represents a started nexus operation will also have an explicit awaitable cancel method.

Reply:
> They need to be awaitable

I don't see how this can be done with cancellation via something you pass in to start often using language-native cancellation mechanisms that are not awaitable.

> This is a lot like how child workflows get cancelled via the external workflow cancel command.

But there is no separate activation for awaiting a child cancel received right? Not sure we have ever had to wait for a cancel to be received by something we've started, be it activities, child workflows, etc.

> cancelling ones you didn't start likely will be supported at some point too.

I think, just as lang would not use cancellation token approach but instead an explicit awaitable cancel call, it would deserve a separate command/activation pair.

Reply:
> I don't see how this can be done with cancellation via something you pass in to start often using language-native cancellation mechanisms that are not awaitable.

They are conceptually awaitable. That doesn't mean the user literally _has_ to wait on them. Like I said you can just track the sequence number and resolution internally for cancel-token-like cases.

> it would deserve a separate command/activation pair.

That is exactly what's happening. You send the `RequestCancelStartedNexusOperation` command, and then you wait for that to be resolved.

> I wonder why we even care about the resolution from lang side then. Not sure it has any lang value. I'd have to see a lang sample that shows where it is even able to be used I think.

This is present in the integ tests - you can see how the Rust SDK is doing this.

@bergundy Suggested I make the cancel explicitly awaitable because of what I mentioned re:
> Support will be added later for ensuring cancel was received

Personally, I don't hugely care. So I'll let him make the case for that if he feels it's important to deliver cancel resolution separately from operation resolution. I would be OK internalizing the after-started cancel, but probably the command will always need to exist (or will have to be reintroduced) when we support cancelling operations you didn't start.

Reply:
To implement `WAIT_CANCELLATION_REQUESTED`, you'll need to notify lang when the cancelation was requested.

Reply:
> To implement WAIT_CANCELLATION_REQUESTED, you'll need to notify lang when the cancelation was requested.

How is this implemented with child workflows today? If we get to external Nexus cancel operation, I agree that has to be a request/response, but for start-in-this-workflow items (children, activities, and Nexus operations), cancellation is a one-way operation. Languages don't even have a way to wait for result.

Can either of y'all show what waiting on a Nexus operation cancel looks like in any of our SDK languages (not to be confused with what waiting on a result looks like). And if 0 SDK languages can support this activation anyways, why would a language ever even handle it? And if 0 SDK languages handle this resulting activation job, why have it effectively as a dead message in the proto?

Reply:
Ah I see what you mean, you don't wait for the cancelation itself in lang, we're expecting core to resolve the nexus operation as canceled with `WAIT_CANCELLATION_REQUESTED`, it's not the result of the cancel command.
I definitely see external nexus operations being a thing and allowing waiting on its cancelation, that may be a separate thing though.

Reply:
So maybe we do want a "Nexus cancellation type"? How does one wait for a Nexus cancellation in Go today? Go child and activity options both have `WaitForCancellation` booleans, should we do the same for Nexus operation options? And for Core-based SDKs, should we have a cancellation type like child and activity have?

Reply:
Yes we need to add a cancellation type for Nexus too. To fully support it, we need to add a history event to notify the SDK that a cancelation request has been delivered (or permanently failed).

Reply:
OK, I'll have operations you started have a non-awaitable cancel operation then I suppose. That's not quite what you mentioned @bergundy when I brought it up in the Nexus channel earlier:

> Make it an awaitable, we are going to add support for waiting for cancellation to be delivered in the future.
It’s basically like child workflow cancellation

But, I can change it back. I might just keep the external cancel command around if we're going to bring that back later.